### PR TITLE
Elasticsearch v2 timestamp support

### PR DESCRIPTION
**Pull request by roysagi**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/19698

## Description
Added support for timestamp in Elasticsearch fetch.

## Does it break backward compatibility?
   - No

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

## Additional changes
Describe additional changes done, for example adding a function to common server.

## Technical writer review
Mention and link to the files that require a technical writer review.
- [ ] [YAML file](link)
- [ ] [CHANGELOG](link)

### CODE REVIEW

File: Integrations/Elasticsearch_v2/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
 ##[Unreleased]  
+Added support for timestamp in Elasticsearch v2.
+   - if the Time Field for fetch is a timestamp simply check the relevant box in the integration setup
+    
+## [19.9.1] - 2019-09-18
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/CHANGELOG.md]: Why did you add a release version?

File: Integrations/Elasticsearch_v2/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,7 @@
 ##[Unreleased]  
+Added support for timestamp in Elasticsearch v2.
+   - if the Time Field for fetch is a timestamp simply check the relevant box in the integration setup
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/CHANGELOG.md]: Don't think that's a good explanation

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 3]
```diff
@@ -38,9 +39,54 @@
...
+    """Gets the last fetch time as a datetime and converts it to the relavent timestamp format.
+
+    Args:
+        last_fetch(datetime): A datetime object setting up the last fetch time
+    """
+    if TIME_METHOD == 'Timestamp-Seconds':
+        return int(last_fetch.timestamp())
+
+    elif TIME_METHOD == 'Timestamp-Milliseconds':
+        return int(last_fetch.timestamp() * 1000)
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: why Don't save the fetch as an int, it will be more accurate this way

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 4]
```diff
@@ -38,9 +39,54 @@
...
+        timestamp_number = float(timestamp_string)
+
+    # convert timestamp to datetime
+    return datetime.fromtimestamp(timestamp_number)
 
 
 def elasticsearch_builder():
+    """Builds an Elasticsearch obj with the necessary credentials, proxy settings and secure connection.
+
+    """
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]:
```suggestion
    """Build an Elasticsearch obj with the necessary credentials, proxy settings and secure connection."""
```

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 5]
```diff
@@ -60,6 +106,14 @@ def elasticsearch_builder():
 
 
 def get_hit_table(hit):
+    """Create context for a single hit in the search.
+
+    Args:
+        hit(Dict): a dictionary representing a single hit in the search.
+
+    Returns:
+        The hit context and the headers of the hit.
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: add type

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 6]
```diff
@@ -76,6 +130,16 @@ def get_hit_table(hit):
...
 def results_to_context(index, query, base_page, size, total_dict, response):
+    """Creates context for the full results of a search.
+
+    Args:
+        index(str): the index in which the search was made.
+        query(str): the query of the search.
+        base_page(int): the base page from which the search is made.
+        size(int): the amount of results to return.
+        total_dict(dict): a dictionary containing the info about thenumber of total results found
+        response(Dict): the raw response of the results.
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: add returns section

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 7]
```diff
@@ -103,6 +167,12 @@ def results_to_context(index, query, base_page, size, total_dict, response):
 
 
 def get_total_results(response_dict):
+    """Creates a dictionary with all for the number of total results found
+
+    Args:
+        response_dict(dict): the raw response from elastic search.
+
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: add returns

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 8]
```diff
@@ -118,6 +188,8 @@ def get_total_results(response_dict):
 
 
 def search_command():
+    """Performs a search in Elasticsearch.
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: the closing """ should be in the same line for one liners, please fix that everywhere

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 9]
```diff
@@ -169,14 +244,20 @@ def fetch_params_check():
     if FETCH_QUERY == '' or FETCH_QUERY is None:
         str_error.append("Query by which to fetch incidents is not configured.")
 
-    if TIME_FORMAT == '' or TIME_FORMAT is None:
+    if (TIME_FORMAT == '' or TIME_FORMAT is None) and TIME_METHOD == 'Simple-Date':
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: shouldn't it be an error for all cases?

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 10]
```diff
@@ -169,14 +244,20 @@ def fetch_params_check():
...
         return_error("Got the following errors in test:\nFetches incidents is enabled.\n" + '\n'.join(str_error))
 
 
 def test_general_query(es):
+    """if is_fetch it ticked, this function runs a generay query to Elasticsearch just to make sure we get a response
+        from the FETCH_INDEX.
+
+    Args:
+        es(Elasticsearch): an Elasticsearch object to which we run the test.
+    """
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: Split it into a one line of explanation for the header
Test executing a fetch query.

then add the note

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 11]
```diff
@@ -217,6 +308,23 @@ def test_fetch_query(es):
...
+    """if is_fetch is ticked and the TIME_METHOD chosen is a type of timestamp - this function checks that
+        the timestamp is in the correct format.
+
+    Args:
+        timestamp(sting): a timestamp string.
+    """
+    timestamp_in_seconds_len = len(str(int(time.time())))
+
+    if TIME_METHOD == 'Timestamp-Seconds' and (len(timestamp) > timestamp_in_seconds_len or not timestamp.isdigit()):
+        return_error(f"The timestamp fetched is not in seconds.\nFetched: {timestamp}")
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: Can you try to identify what it is?

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 12]
```diff
@@ -283,10 +406,28 @@ def incident_label_maker(source):
...
+    """Converts the current results into incidents.
+
+    Args:
+        response(dict): the raw search results from Elasticsearch.
+        current_fetch(datetime): the date of the last fetch before this fetch.
+        last_fetch(datetime): the date of the last fetch before this fetch - this will hold the last date of the
+            incident brought by this fetch.
+
+    Returns:
+        the inicidents and the date of the last incident brought by this fetch.
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: type

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py (hunk 4 above)
Comment (roysagi) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: done

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py (hunk 5 above)
Comment (roysagi) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: done

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py (hunk 6 above)
Comment (roysagi) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: done

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py (hunk 7 above)
Comment (roysagi) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: done

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py (hunk 8 above)
Comment (roysagi) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: done - everywhere

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py (hunk 9 above)
Comment (roysagi) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: no, in case it is a timestamp the TIME_FORMAT is not necessary because the TIME_FORMAT is predefined (lines 45-46)

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py (hunk 11 above)
Comment (roysagi) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: done (same in milliseconds part obviously)

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py (hunk 12 above)
Comment (roysagi) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: done

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py (hunk 3 above)
Comment (roysagi) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: I didn't want to change the entire logic of the fetch. So in this function after getting the last fetch time I convert it to the correct timestamp format.

I can change the integration to work with a timestamp throughout the fetch process but I thought it would be better to be less intrusive and change the overall logic.

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py (hunk 10 above)
Comment (roysagi) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: done - in all test functions

File: Integrations/Elasticsearch_v2/CHANGELOG.md (hunk 1 above)
Comment (roysagi) [reply Integrations/Elasticsearch_v2/CHANGELOG.md]: wasn't I supposed to? - the notes below were already released.

File: Integrations/Elasticsearch_v2/CHANGELOG.md (hunk 2 above)
Comment (roysagi) [reply Integrations/Elasticsearch_v2/CHANGELOG.md]: removed it

File: Integrations/Elasticsearch_v2/CHANGELOG.md (hunk 1 above)
Comment (yardensade) [reply Integrations/Elasticsearch_v2/CHANGELOG.md]: nope we didn't, this will require manual fixing for now, the actual fix for the issue will be in the next clear. (today)

**Comment by roysagi** [comment]
@kirbles19 - If you could please take a look at these files:

https://github.com/demisto/content/blob/elastic_timestamp_support/Integrations/Elasticsearch_v2/CHANGELOG.md

https://github.com/demisto/content/blob/elastic_timestamp_support/Integrations/Elasticsearch_v2/Elasticsearch_v2.yml

https://github.com/demisto/content/blob/elastic_timestamp_support/Integrations/Elasticsearch_v2/Elasticsearch_v2_description.md

**Comment by ronykoz** [comment]
@roysagi i don't see those UTs we talked about

**Comment by roysagi** [comment]
> @roysagi i don't see those UTs we talked about

for some reason it didn't update it... I commited again now it's up to date

**Comment by kirbles19** [comment]
@roysagi Done.
